src/utils/eis_utils/src/eis_utils.c

```diff
@@ -468,208 +550,241 @@ EISUtilityResult RequestConnectionStringFromEISWithExpiry(
+
+        const JSON_Object* authJson =
+            json_value_get_object(json_object_get_value(specJson, EIS_IDENTITY_RESP_AUTH_FIELD));
 
-        if (certResponseJson == NULL)
+        if (authJson == NULL)
         {
             result.err = EISErr_InvalidJsonRespErr;
-            result.service = EISService_CertService;
+            result.service = EISService_IdentityService;
             goto done;
         }
 
-        const JSON_Object* certResponseJsonObj = json_value_get_object(certResponseJson);
-
-        const char* certificateStr = json_object_get_string(certResponseJsonObj, EIS_CERT_RESP_PEM);
+        const char* authTypeStr = json_object_get_string(authJson, EIS_IDENTITY_RESP_AUTH_TYPE_FIELD);
 
-        if (certificateStr == NULL)
+        if (authTypeStr == NULL)
         {
             result.err = EISErr_InvalidJsonRespErr;
-            result.service = EISService_CertService;
+            result.service = EISService_IdentityService;
             goto done;
         }
 
-        if (mallocAndStrcpy_s(&certString, certificateStr) != 0)
+        const char* keyHandle = json_object_get_string(authJson, EIS_IDENTITY_RESP_AUTH_KEYHANDLE_FIELD);
+
+        if (keyHandle == NULL)
         {
-            result.err = EISErr_ContentAllocErr;
+            result.err = EISErr_InvalidJsonRespErr;
+            result.service = EISService_KeyService;
             goto done;
         }
 
-        result.err =
-            BuildSasCertConnectionString(hubName, deviceId, moduleId, connType, gatewayHostName, &connectionStr);
+        if (strcmp(authTypeStr, "sas") == 0)
+        {
+            authType = ADUC_AuthType_SASToken;
+
+            result = BuildSharedAccessSignature(
+                resourceUri, keyHandle, expirySecsSinceEpoch, timeoutMS, &sharedSignatureStr);
+
+            if (result.err != EISErr_Ok)
+            {
+                goto done;
+            }
+
+            result.err = BuildSasTokenConnectionString(
+                hubName, deviceId, moduleId, connType, sharedSignatureStr, gatewayHostName, &connectionStr);
 
-        if (result.err != EISErr_Ok)
+            if (result.err != EISErr_Ok)
+            {
+                goto done;
+            }
+        }
+        else if (strcmp(authTypeStr, "x509") == 0)
         {
+            authType = ADUC_AuthType_SASCert;
+
+            if (mallocAndStrcpy_s(&keyHandlePtr, keyHandle) != 0)
+            {
+                result.err = EISErr_ContentAllocErr;
+                goto done;
+            }
+
+            const char* certId = json_object_get_string(authJson, EIS_IDENTITY_RESP_AUTH_CERTID_FIELD);
+
+            if (certId == NULL)
+            {
+                result.err = EISErr_InvalidJsonRespErr;
+                result.service = EISService_IdentityService;
+                goto done;
+            }
+
+            EISErr certResult = RequestCertificateFromEIS(certId, timeoutMS, &certResponseStr);
```

Comment:
The UID of this process needs to be the same as the one in the principle (ais_user I believe) in order to make this call. 

It looks like the file reader may be a separate user?

Reply:
yeah it is a separate user... oh no